lib/teiserver/account.ex

```diff
@@ -1887,42 +1887,73 @@ defmodule Teiserver.Account do
 
   """
   def create_friend_request(attrs \\ %{}) do
-    result =
-      %FriendRequest{}
-      |> FriendRequest.changeset(attrs)
-      |> Repo.insert()
-
-    case result do
-      {:ok, friend_request} ->
-        PubSub.broadcast(
-          Teiserver.PubSub,
-          "account_user_relationships:#{friend_request.to_user_id}",
-          %{
-            channel: "account_user_relationships:#{friend_request.to_user_id}",
-            event: :new_incoming_friend_request,
-            userid: friend_request.to_user_id,
-            from_id: friend_request.from_user_id
-          }
-        )
-
-        Teiserver.cache_delete(:account_incoming_friend_request_cache, friend_request.to_user_id)
-        Teiserver.cache_delete(:account_outgoing_friend_request_cache, friend_request.to_user_id)
-
-      _ ->
-        :ok
-    end
-
-    result
+    # NOTE: All call sites must now handle {:error, reason} returns from create_friend_request/2
+    from_user_id = Map.get(attrs, :from_user_id)
+    to_user_id = Map.get(attrs, :to_user_id)
+    create_friend_request(from_user_id, to_user_id)
   end
 
   def create_friend_request(from_user_id, to_user_id) do
     if from_user_id == to_user_id do
       {:error, "Cannot add yourself as a friend"}
     else
-      create_friend_request(%{
-        from_user_id: from_user_id,
-        to_user_id: to_user_id
-      })
+      {allowed, reason} = can_send_friend_request_with_reason?(from_user_id, to_user_id)
+
+      if not allowed do
+        {:error, reason}
+      else
+        Repo.transaction(fn ->
+          existing =
+            Repo.one(
+              from fr in FriendRequest,
+                where:
+                  (fr.from_user_id == ^from_user_id and fr.to_user_id == ^to_user_id) or
+                    (fr.from_user_id == ^to_user_id and fr.to_user_id == ^from_user_id)
+            )
+
+          if existing do
+            Repo.rollback("Pending request exists")
+          else
```

Comment:
`can_send_friend_request_with_reason?` already checks if an outgoing friend request exists.

Additionally, it would be nice if an incoming friend request could be auto accepted if you are trying to send a new request to the user that sent you a friend request.
So for example,
- User A sends friend request to user B
- User B attempts to send friend request to user A
- User A's friend request is accepted